### DISCUSSION TITLE
Parsing and interacting with Scheme AST

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#90] [Discussions]
@leppie,

I need to get at the AST of Scheme forms for pretty printing and compacting Scheme code. By pretty printing I mean properly formatting the code, like in an editor. By compacting I mean taking a multiline block of code and removing unnecessary whitespace including line feeds so the entire expression can be converted to a one-line text string. The operation should work in either direction, and any syntax errors should be available for me to present to the user as appropriate.

As a simple but concrete example, I'd like to be able to go from this:

```scheme
(lambda (a b)
  (if (> a b)
      "a is greater"
      "b is greater"))
```
to this:

```
"(lambda (a b) (if (> a b) ""a is greater"" ""b is greater""))"
```

and vice versa.

I'm sure properly formatting Scheme code is the subject of flame wars but I just one one way that looks good enough and is, more importantly, consistent.

I started playing around with the Parser and Scanner classes in IronScheme.Compiler but don't really know how to use them to get what I want (probably what I want is somewhere up in the DLR interfaces, perhaps).  Any pointers how to get the AST and play with it for my purposes, or where to look for existing examples, appreciated.

**Comment by leppie** [comment]
There is a basic version of `pretty-print` in IronScheme. It is not really meant for formatting code, but it can do a little.

Example
```
> (pretty-print
. '(lambda (a b)
.   (if (> a b)
.       "a is greater"
.       "b is greater"))
. )
(lambda (a b) (if (> a b) "a is greater" "b is greater"))
58
> (pretty-width 32)
> (pretty-print
. '(lambda (a b)
.   (if (> a b)
.       "a is greater"
.       "b is greater"))
. )
(lambda (a b)
  (if (> a b)
    "a is greater"
    "b is greater"))
21
>
```

Code is here if you want to look: https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/build/pretty-print.sls

PS: Going into the AST you going to end up with a macro expanded code with will be nigh impossible to reconstruct.

**Reply by bcalco** [reply]
OK so if I want `pretty-print` to turn a multi-line form into a single line form I need to know how many characters it contains and make sure `pretty-width` is set to a number bigger than that, otherwise use some reasonably short width (I guess based on the length of the longest line) to force it to format it "pretty"... this might be workable.

I totally get the point about macros. Perhaps access to raw tokens instead of the AST might be sufficient. But I'll see if I can make `pretty-print` work for now, as a workaround.

**Reply by bcalco** [reply]
I should have been clearer that I need this functionality on the C# side. What is coming back is the width of the longest line, not the actual text lines. While the number is interesting, what I really want is the pretty printed string. Calling "Eval()" on

```csharp
      string command = String.Format (myPrettyPrintCommand, maxWidth, aSchemeForm );
      object evalResult = command.Eval ();
```

where `myPrettyPrintCommand` is defined as

```csharp
const string myPrettyPrintCommand = "(pretty-width {0}) (pretty-print `{1})";
```

gives me an integer result.

**Reply by bcalco** [reply]
So looking at the code in `pretty-print.sls` it seems I can pass an alternate `port`  in the optional `opt` parameter to `(pretty-print obj . opt)` form, otherwise it grabs `(current-output-port)`... So perhaps I can write a helper function that creates a text stream that captures the output, and have that returned from my custom function... which I can then call via a constructed `Eval()` command.

BTW, I assume everything in the build directory including `pretty-print.sls` is what gets compiled into ironscheme.boot.dll, is that correct?

How might I create my own separate `myapp.boot.dll` for precompiled Scheme code I want to "always be there" on startup? (Or is the assumption here I'd include an `init.ss` or boot script on startup for that purpose?

**Reply by leppie** [reply]
> BTW, I assume everything in the build directory including `pretty-print.sls` is what gets compiled into ironscheme.boot.dll, is that correct?

That is correct.

**Reply by leppie** [reply]
> How might I create my own separate `myapp.boot.dll` for precompiled Scheme code I want to "always be there" on startup? (Or is the assumption here I'd include an `init.ss` or boot script on startup for that purpose?

The easiest would be to just call `(import (your-init-lib))`. I think there might be a `init.ss` file that is called on the REPL, but not sure if that is hooked up for C# Eval. 

It really depends on your scenario.

**Comment by lassik** [comment]
@bcalco You might be able to hack the [laesare](https://gitlab.com/weinholt/laesare) library to do what you need.

**Reply by lassik** [reply]
See also https://github.com/lispunion/code-formatter/issues/5

**Reply by bcalco** [reply]
Thanks for the references. I'm going to see if including `laesare` helps, since `pretty-print` returns an integer and not the pretty printed string. First I'll see if I can create a version of `pretty-print` that returns what I want, of course.

The Ocaml implementation pointers in [Strictly Pretty](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.34.2200&rep=rep1&type=pdf) should be translatable to F#, if I really need to implement pretty printing without the aid of built-in functionality in the IronScheme C# or Scheme code.

**Comment by bcalco** [comment]
OK, I think I solved my problem.

Given:

```csharp
    const string myPrettyPrintCommand = "(let ((p (open-output-string))) "
                                      +   "(begin (pretty-width {0}) "
                                      +          "(pretty-print `{1} p) " 
                                      +          "(get-output-string p)))"
                                      ;
```
This:

```csharp
    public string PrettyPrint ( string aSchemeForm )
    {      
      string result = "";
      int maxWidth = 32; // default

      if ( aSchemeForm.Length < maxWidth )
        maxWidth = aSchemeForm.Length - 1;

      string command = String.Format (myPrettyPrintCommand, maxWidth, aSchemeForm);
      object evalResult = command.Eval ();
      result = evalResult.ToString ();
      return result;
    }
```

Seems to let me work with the built-in pretty print functionality. My job is now to know when to set pretty-width to what, variously to expand or compact the text. 

Digging around `ports.sls` was the key. I just knew there had to be a text reader port somewhere. :) 

What wasn't intuitive was the fact that the `pretty-print` function seemed to want a list (hence the `pair?` guard) for the port provided in the `opt` variable, so I spent a lot of time getting ` not textual-output-port"` errors on `'(p)` till I said screw it let's just pass the port in, assuming whatever it creates is a pair, and it started working. At least, I was getting a formatted string back. So this should get me going.
